auth_backend/backends.py

```diff
@@ -34,10 +34,25 @@ def authenticate(self, email=None, username=None, password=None, **kwargs):
         if strategy:
             payload['strategy'] = strategy
 
+        auth_api_client = AuthApiClient(cas_credentials)
         status, data = auth_api_client.call('sessions', 'POST', payload)
 
         if status not in (200, 404,):
             raise CASUnexpectedStatusCode(status, data)
 
         if status == 200:
-            return user
+            if existing_user:
+                existing_user.cas_credentials = cas_credentials
+            else:
+                try:
+                    # Do not on save sync to CAS, as we just got the user's
+                    # data from CAS, and nothing has changed in the interim
+                    pre_save.disconnect(save_user_to_cas, sender=KagisoUser)
+                    existing_user = KagisoUser(cas_credentials)
+                    existing_user.set_password(password)
+                    existing_user.create_from_cas_data(data)
+                    existing_user.save()
+                finally:
+                    pre_save.connect(save_user_to_cas, sender=KagisoUser)
```

Comment:
how bout in your `build_from_cas_data` function set a variable `built_from_cas`

then in your `save_user_to_cas` function.

```
if built_from_cas:
    return
```

i think its not linear and it feels like you have to work around the fact you already set up the receivers in your models.py

my suggestion allows you to remove the `disconnect` and `connect` calls